### PR TITLE
Validate custom object types and properties in patterns

### LINKED ISSUE #23: Validate that indicator patterns use valid observable types and property names
Issues like oasis-open/cti-stix2-json-schemas#55 are caused because we don't validate the contents of Indicator pattern expressions, just the syntax. We should be able to combine the [stix2patterns.inspector module](https://github.com/oasis-open/cti-pattern-validator/blob/master/stix2patterns/inspector.py) with the data in `enums.py` to check whether object types and properties (including via referenced objects) are valid.

For custom types and properties, we can follow the same rules as the rest of the validator:
- allow them regardless if they follow the correct naming conventions, and 
- raise an error if they don't, which can be ignored/silenced.

To scope this, I'll assume the following are out of scope for now:
- Checking if the values are the right type (since this isn't yet tracked in enums.py)
- Allowing users to specify names of properties of custom objects in or nested property names in custom properties. So basically, as soon as a (valid) custom object name or property name is encountered, checking that component of the pattern stops.

### PR DESCRIPTION
**Pull request by clenk**
Does not check properties of extensions or embedded objects. That will be possible once `stix2patterns` has been updated so the `inspect()` method gives more granular details.

Note that this PR removes some check codes; `custom-property-prefix`, `custom-observable-object-prefix,` `custom-object-extension-prefix`, `custom-observable-properties-prefix`, and their `-lax` siblings have been replaced with `custom-prefix` and `custom-prefix-lax`.

Fix #23.

### CODE REVIEW

File: stix2validator/shoulds.py [hunk 1]
```diff
@@ -22,6 +22,37 @@
 from .output import info
 
 
+def custom_prefix_strict(instance):
+    """Ensure custom content follows strict naming style conventions.
+    """
+    for error in custom_object_prefix_strict(instance):
```
Comment (gtback) [inline stix2validator/shoulds.py]: can probably use `itertools.chain` to make this easier

File: stix2validator/musts.py [hunk 2]
```diff
@@ -327,18 +328,71 @@ def types_strict(instance):
...
+    p = Pattern(pattern)
+    inspection = p.inspect().comparisons
+    for objtype in inspection:
+        # Check observable object types
+        if objtype in enums.OBSERVABLE_TYPES:
+            pass
+        elif options.strict_types:
+            yield PatternError("'%s' is not a valid STIX observable type"
+                               % objtype, instance['id'])
+        elif (not re.match('^\\-?[a-z0-9]+(-[a-z0-9]+)*\\-?$', objtype) or
```
Comment (gtback) [inline stix2validator/musts.py]: Consider pulling out these (and other) regular expressions into pre-compiled RE objects. Giving them a name makes them more descriptive, as well as making them easier to test independently (don't necessarily worry about adding those tests now).

File: stix2validator/musts.py [hunk 3]
```diff
@@ -327,18 +328,71 @@ def types_strict(instance):
...
+              len(objtype) < 3 or len(objtype) > 250):
+            yield PatternError("'%s' is not a valid observable type name"
+                               % objtype, instance['id'])
+        elif (all(x not in options.disabled for x in ['all', 'format-checks', 'custom-prefix']) and
+              not re.match("^x\-.+\-.+$", objtype)):
+            yield PatternError("Custom Observable Object type '%s' should start "
+                               "with 'x-' followed by a source unique identifier "
+                               "(like a domain name with dots replaced by "
+                               "dashes), a dash and then the name"
```
Comment (gtback) [inline stix2validator/musts.py]: s/dash/hyphen/

File: stix2validator/musts.py [hunk 4]
```diff
@@ -327,18 +328,71 @@ def types_strict(instance):
...
+                                   % (prop, objtype), instance['id'])
+            elif not re.match('^[a-z0-9_]{3,250}$', prop):
+                yield PatternError("'%s' is not a valid observable property name"
+                                   % prop, instance['id'])
+            elif (all(x not in options.disabled for x in ['all', 'format-checks', 'custom-prefix']) and
+                  not re.match("^x_.+_.+$", prop)):
+                yield PatternError("Cyber Observable Object custom property '%s' "
+                                   "should start with 'x_' followed by a source "
+                                   "unique identifier (like a domain name with "
+                                   "dots replaced by dashes), a dash and then the"
```
Comment (gtback) [inline stix2validator/musts.py]: s/dash/underscore/

File: stix2validator/test/indicator_tests.py [hunk 5]
```diff
@@ -140,6 +140,31 @@ def test_invalid_pattern(self):
         indicator['pattern'] = """[file:hashes."SHA-256" = 'aec070645fe53ee3b3763059376134f058cc337247c978add178b6ccdfb0019f']"""
         self.assertFalseWithOptions(json.dumps(indicator))
 
+    def test_pattern_custom_invalid_format(self):
+        indicator = copy.deepcopy(self.valid_indicator)
```
Comment (gtback) [inline stix2validator/test/indicator_tests.py]: I've been curious. Why do we deep-copy the indicator before each test? Are there tests that mutate the object?

Comment (clenk) [inline stix2validator/test/indicator_tests.py]: Yes, several tests modify the object; for instance several in this file modify the indicator's pattern. I suppose it wouldn't be any different if we just did json.loads() in each test.

Comment (gtback) [inline stix2validator/test/indicator_tests.py]: Ok, don't worry about it. I don't know why, but seeing `deepcopy` in code always looks ugly to me, for no good reason.

**Pull request merged**